### PR TITLE
David

### PR DESCRIPTION
**Pull request by dreeds**
Implemented save/load

### CODE REVIEW

File: src/edu/psu/sweng500/team8/coreDataStructures/SavePackage.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package edu.psu.sweng500.team8.coreDataStructures;
+
+import java.io.Serializable;
+
+import javax.swing.JTextField;
+
+public class SavePackage  implements Serializable {
+	private CellGrid m_cellGrid;
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/coreDataStructures/SavePackage.java]: The java convention does not use m_ . I used it out of habit from c++ but later learned this is not the standard

File: src/edu/psu/sweng500/team8/gui/GridPanel.java [hunk 2]
```diff
@@ -58,8 +64,14 @@ public void populatePanel(CellGrid grid, GameSession gameSession) {
                 cellTextBox.setEditable(true); //Need to set Editable first to set the number
                 if (cell.hasNumber()) {
                 	cellTextBox.setText(Integer.toString(cell.getNumber()));
-
-                    markGivenCell(cell);
+	                
+                	if(cell.getType() == ValueType.Given){
+	                	markGivenCell(cell);
+	                	
+	                	CellCoordinates coordinates = cell.getCoordinates();
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/GridPanel.java]: markGivenCell already does this. You can remove this

File: src/edu/psu/sweng500/team8/gui/GridPanel.java [hunk 3]
```diff
@@ -73,6 +85,23 @@ public void populatePanel(CellGrid grid, GameSession gameSession) {
             }
         }
     }
+    
+    public void remarkGivenCells(CellGrid grid){
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/GridPanel.java]: Why is this necessary? It is only called in the Save button handler, but that shouldn't clear the highlights. In fact I commented it out and it didn't seem to make a difference. suggest removing this method

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 4]
```diff
@@ -66,10 +80,22 @@ public void cellChanged(Cell cell, int newNumber) {
 		clearMessage();
 		this.gameBoard.clearHighlightedIncorrectCells();
 		
+		this.board = this.gameSession.getGameBoard();
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: This shouldn't be necessary. I don't have a problem with adding it, but then we should use this consistently throughout.

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 5]
```diff
@@ -66,10 +80,22 @@ public void cellChanged(Cell cell, int newNumber) {
...
+		
+		
+		if (!board.hasOpenCells()) {
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: This is redundant -- gameIsComplete() does the same thing. Looks like old code got merged back in.

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 6]
```diff
@@ -134,6 +160,16 @@ public void actionPerformed(java.awt.event.ActionEvent evt) {
         jLabel2.setText("Difficulty");
 
         jButton10.setText("Save");
+        jButton10.addActionListener(new java.awt.event.ActionListener() {
+            public void actionPerformed(java.awt.event.ActionEvent evt) {
+            	try {
+					jButton10ActionPerformed(evt);
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: Fix tab formatting

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 7]
```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
...
 		// Create a file chooser
+		
 		final JFileChooser fc = new JFileChooser();
-		// In response to a button click:
+		fc.setFileFilter(new FileFilter()
+        {
+           @Override
+           public boolean accept(File file)
+           {
+              return (file.getName().toUpperCase().endsWith(".SUDOKU") || file.isDirectory());
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: directories are accepted? What happens in that case?

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 8]
```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
...
+              return "Sudoku files";
+           }
+        });
+		
+		fc.setAcceptAllFileFilterUsed(false);
 		int returnVal = fc.showOpenDialog(getComponent(0));
 
+		if(returnVal == JFileChooser.APPROVE_OPTION){
+		
+			SavePackage savePackage = null;
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: Fix tab indending

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 9]
```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
...
+	      try
+	      {
+	         FileInputStream fileIn = new FileInputStream(fc.getSelectedFile().getAbsolutePath());
+	         ObjectInputStream in = new ObjectInputStream(fileIn);
+	         savePackage = (SavePackage) in.readObject();
+	         in.close();
+	         fileIn.close();
+	         this.loadSession(savePackage);
+	      }catch(IOException i)
+	      {
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: Inconsistent brace style -- should be on the end of the preceding line. Same below.

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 10]
```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
...
+	         i.printStackTrace();
+	         return;
+	      }catch(ClassNotFoundException c)
+	      {
+	         System.out.println("Employee class not found");
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: Employee? Copy/paste?

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 11]
```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
...
+	         FileInputStream fileIn = new FileInputStream(fc.getSelectedFile().getAbsolutePath());
+	         ObjectInputStream in = new ObjectInputStream(fileIn);
+	         savePackage = (SavePackage) in.readObject();
+	         in.close();
+	         fileIn.close();
+	         this.loadSession(savePackage);
+	      }catch(IOException i)
+	      {
+	         i.printStackTrace();
+	         return;
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: return statement is redundant. Same below

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 12]
```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
...
+	         c.printStackTrace();
+	         return;
+	      }
+		}
+		
 	}// GEN-LAST:event_jButton11ActionPerformed
+	
 
+	
+    private void jButton10ActionPerformed(java.awt.event.ActionEvent evt) throws FileNotFoundException {
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: We should rename these buttons in the designer. "jButton#" is a bit confusing

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 13]
```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
...
+
+				File f = fc.getSelectedFile();
+				String path = f.getAbsolutePath();
+				
+				if(f.exists()){
+		            int result = JOptionPane.showConfirmDialog(this,"The file exists, overwrite?","Existing file",JOptionPane.YES_NO_CANCEL_OPTION);
+		            switch(result){
+		                case JOptionPane.YES_OPTION:
+		                    fc.approveSelection();
+		                   // savePuzzle(path);
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: I found this while testing -- the overwrite option does not work. (BUG)

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 14]
```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
...
+		                    return;
+		                case JOptionPane.NO_OPTION:
+		                    return;
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: Shouldn't these call cancelSeletion() ?

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 15]
```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
...
+    private void savePuzzle(String path){
+		try
+	      {
+		
+			if(!path.toLowerCase().endsWith(".sudoku")){
+				path = path + ".sudoku";
+			}
+			
+			SavePackage savePackage = new SavePackage();
+			savePackage.setCellGrid(this.gameSession.getGameBoard().getCellGrid());
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: Either use this.gameSession.getGameBoard() throughout and remove this.board OR use this.board consistently

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 16]
```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
...
+			savePackage.setPuzzle(this.gameSession.getGameBoard().getCurrentPuzzle());
+			savePackage.setControlGrid(this.gameBoard.getControlGrid());
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: Since everything being saved is in the board, would it make sense to just save the board?

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 17]
```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
...
+
+	      }catch(IOException i)
+	      {
+	          i.printStackTrace();
+	      }
+
+    }
+	
+	private void loadSession(SavePackage savePackage){
+		//this.gameBoard.populatePanel(gameSession.getGameBoard().getCellGrid(), gameSession);
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: Remove commented-out code

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 18]
```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
...
+		
+		Puzzle puzzle = savePackage.getPuzzle();
+		
+		setMessage("");
+		DifficultyLevel difficulty;
+		if (puzzle.getDifficulty() == DifficultyLevel.Easy){
+			radEasy.setSelected(true);			
+		}
+
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: Remove extra whitespace between if/else statements

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 19]
```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
...
+		}
+		
+		else{
+			radHard.setSelected(true);
+		}
+				
+		this.gameSession = new GameSession(puzzle);
+		this.gameSession.subscribeForCellChanges(this);
+		
+		GridPanel gameBoard = new GridPanel(savePackage.getControlGrid());
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: suggest renaming this so it is not confused with this.gameBoard

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java [hunk 20]
```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
...
+		//this.gameBoard = gameBoard;
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: Remove commented-out code

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 19 above)
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: This is loaded then thrown away...? I don't think it is necessary to save the JTextField[][] in the save package

File: src/edu/psu/sweng500/team8/play/GameSession.java [hunk 21]
```diff
@@ -14,8 +15,9 @@
  * If we are going to use Board class for operations, this class is not needed and we can move operations from this class to Board class
  * How are we going to response back to UI when redo/undo happens? Are we going to simply update board/grid and let UI call refresh?
  */
-public class GameSession {
+public class GameSession  implements Serializable{
 	
+	private static final long serialVersionUID = 7878878;
```
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/play/GameSession.java]: Suggest using a meaningful number, such as today's date: 071615

File: src/edu/psu/sweng500/team8/coreDataStructures/SavePackage.java (hunk 1 above)
Comment (dreeds) [inline src/edu/psu/sweng500/team8/coreDataStructures/SavePackage.java]: If you think that we should refactor them all, I volunteer

File: src/edu/psu/sweng500/team8/gui/GridPanel.java (hunk 3 above)
Comment (dreeds) [inline src/edu/psu/sweng500/team8/gui/GridPanel.java]: I don’t understand why it should make a difference, but without it, when I save, the highlights disappear.

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 4 above)
Comment (dreeds) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: I removed the ones that I had added

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 5 above)
Comment (dreeds) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: The above is not my code

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 6 above)
Comment (dreeds) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: They are formatted correctly on my machine

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 7 above)
Comment (dreeds) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: It shows the directories which can be opened, so we can operate on the files inside. If we don't accept directories only the files are shown.

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 8 above)
Comment (dreeds) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: fixed.

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 9 above)
Comment (dreeds) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: fixed as many as I could find. C# habits.

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 16 above)
Comment (dreeds) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: The board has swing components which aren't serializable.

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 15 above)
Comment (dreeds) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: removed this.board; removed the variable board

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 17 above)
Comment (dreeds) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: removed.

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 18 above)
Comment (dreeds) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: removed

File: src/edu/psu/sweng500/team8/coreDataStructures/SavePackage.java (hunk 1 above)
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/coreDataStructures/SavePackage.java]: At least for new classes, please do.

File: src/edu/psu/sweng500/team8/gui/GridPanel.java (hunk 3 above)
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/GridPanel.java]: Hmmmm I'm not seeing that. If it doesn't work, there must be some other explanation. Can you debug it and find out?

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 5 above)
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: That's what I'm saying -- it was old code that was refactored into gameIsComplete(). It must have been re-added during the merge. The fact that it's showing as a diff in this review means your branch is re-adding it. Please remove it before merging this.

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 7 above)
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: Ok

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 9 above)
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: I hear ya :) I'm not a fan of the java style either.

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 16 above)
Comment (jernelson7) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: I mean this.gameSession.getGameBoard(). It is a coreDataStructures.Board which does not have any Swing components.

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 20 above)
Comment (dreeds) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: removed

File: src/edu/psu/sweng500/team8/gui/SudokuGUI.java (hunk 19 above)
Comment (dreeds) [inline src/edu/psu/sweng500/team8/gui/SudokuGUI.java]: changed.
